### PR TITLE
feat: add fusco support

### PR DESCRIPTION
**Pull request by zoran-zhu**
### Description
This PR introduces FUSCO support to RLinf to accelerate MoE (Mixture-of-Experts) communication.

Specifically, it integrates FUSCO as an alternative communication backend for Megatron MoE dispatch and combine operations. When enabled, the FUSCO-based implementation replaces the original communication path while preserving the same semantic behavior.

The feature is optional and automatically falls back to the default implementation when FUSCO is not available.

### How has this been tested?
Tested by enabling FUSCO support and running end-to-end MoE tests.
#### 1. Install FUSCO
```bash
# Clone and install the FUSCO package:
git clone https://github.com/infinigence/FUSCO
cd FUSCO
python setup.py install
mkdir -p lib
curl -L -o lib/libfusco.so https://ghfast.top/https://github.com/infinigence/FUSCO/releases/download/v0.1/libfusco.so
```
#### 2. Run e2e test
```bash
FUSCO_SO_PATH=/path/to/libfusco.so \
REPO_PATH=/path/to/RLinf/ \
bash tests/e2e_tests/reasoning/run.sh \
qwen3-moe-2.5b-collocated-mg-sgl-ep-fusco-test
```

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Documentation update (Document-only update)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

### Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: rlinf/workers/actor/megatron_actor_worker.py [hunk 1]
```diff
@@ -452,6 +454,9 @@ def forward_output_and_loss_func(dataloader_iter, model):
             input_ids = batch["input_ids"]
             attention_mask = batch["attention_mask"]
             position_ids = batch["position_ids"]
+            padding_seqlen = None
+            if self.variable_seq_lengths is False:
+                padding_seqlen = self.encoder_seq_length
```
Comment (qurakchin) [inline rlinf/workers/actor/megatron_actor_worker.py]: when dynamic_batch_size is enabled, the padding seq len is self.max_tokens_per_mbs. please add a test with dynamic_batch_size enabled.

File: tests/e2e_tests/reasoning/qwen3-moe-2.5b-collocated-mg-sgl-ep-fusco-test.yaml [hunk 2]
```diff
@@ -0,0 +1,298 @@
...
+  max_epochs: 1
+  max_steps: 5
+
+  val_check_interval: 1
+  save_interval: -1
+
+  seq_length: 1024
+
+  enable_dynamic_batch_size: True
+  max_tokens_per_mbs: 1024
```
Comment (qurakchin) [inline tests/e2e_tests/reasoning/qwen3-moe-2.5b-collocated-mg-sgl-ep-fusco-test.yaml:31]: add a test with enable_dynamic_batch_size and max_tokens_per_mbs > 1024

File: rlinf/hybrid_engines/megatron/utils.py [hunk 3]
```diff
@@ -50,6 +50,10 @@ def preprocess_packed_seqs(
 
     shape = list(input_ids.shape[1:])
     shape[0] = seqlens_in_batch_padded.sum().item() // cp_size
+    if padding_seqlen is None:
+        shape[0] = seqlens_in_batch_padded.sum().item() // cp_size
```
Comment (qurakchin) [inline rlinf/hybrid_engines/megatron/utils.py:58]: add a test with cp enabled

File: rlinf/hybrid_engines/megatron/megatron_model_manager.py [hunk 4]
```diff
@@ -141,6 +157,29 @@ def __init__(self, cfg: DictConfig):
         # In AUTO mode, the actor will occupy all GPUs for initialization, but not all Megatron processes will be in the running state.
         self.is_running = True
 
+        # Patch Megatron MoE token dispatcher if FUSCO is available and conditions are met
+        self.patch_megatron_moe_dispatcher()
+
+    def patch_megatron_moe_dispatcher(self):
+        if (
+            HAVE_FUSCO 
+            and self._cfg.model.moe_token_dispatcher_type == "alltoall"
```
Comment (FxxxxU) [inline rlinf/hybrid_engines/megatron/megatron_model_manager.py]: Please, use assert  to check the special parameter. It's bettert to move this check to config.py file.

**Review by FxxxxU — APPROVED** [review]

File: tests/e2e_tests/reasoning/qwen3-moe-2.5b-collocated-mg-sgl-ep-fusco-test.yaml [hunk 5]
```diff
@@ -0,0 +1,298 @@
...
+runner:
+  task_type: reasoning
+  logger:
+    log_path: /workspace/results/
+    project_name: rlinf
+    experiment_name: "ci-test"
+    logger_backends: ["tensorboard"] # wandb, swanlab
+
+  max_epochs: 1
+  max_steps: 5
```
Comment (andylin-hao) [inline tests/e2e_tests/reasoning/qwen3-moe-2.5b-collocated-mg-sgl-ep-fusco-test.yaml]: Maybe reduce the max_steps to 1 to save some test time
